Add simplified model manager install API to InvocationContext

lstein opened the pull request:
## Summary

This three two model manager-related methods to the InvocationContext uniform API. They are accessible via `context.models.*`:

1. **`load_local_model(model_path: Path, loader: Optional[Callable[[Path], AnyModel]] = None) -> LoadedModelWithoutConfig`**

*Load the model located at the indicated path.*

This will load a local model (.safetensors, .ckpt or diffusers directory) into the model manager RAM cache and return its `LoadedModelWithoutConfig`. If the optional loader argument is provided, the loader will be invoked to load the model into memory. Otherwise the method will call  `safetensors.torch.load_file()`  `torch.load()` (with a pickle scan), or `from_pretrained()` as appropriate to the path type.

Be aware that the `LoadedModelWithoutConfig` object differs from `LoadedModel` by having no `config` attribute.

Here is an example of usage:

```
def invoke(self, context: InvocatinContext) -> ImageOutput:
       model_path = Path('/opt/models/RealESRGAN_x4plus.pth')
       loadnet = context.models.load_local_model(model_path)
       with loadnet as loadnet_model:
             upscaler = RealESRGAN(loadnet=loadnet_model,...)
```

---

2. **`load_remote_model(source: str | AnyHttpUrl, loader: Optional[Callable[[Path], AnyModel]] = None) -> LoadedModelWithoutConfig`**

*Load the model located at the indicated URL or repo_id.*

This is similar to `load_local_model()` but it accepts either a HugginFace repo_id (as a string), or a URL. The model's file(s) will be downloaded to `models/.download_cache` and then loaded, returning a 

```
def invoke(self, context: InvocatinContext) -> ImageOutput:
       model_url = 'https://github.com/xinntao/Real-ESRGAN/releases/download/v0.1.0/RealESRGAN_x4plus.pth'
       loadnet = context.models.load_remote_model(model_url)
       with loadnet as loadnet_model:
             upscaler = RealESRGAN(loadnet=loadnet_model,...)
```
---

3. **`download_and_cache_model( source: str | AnyHttpUrl, access_token: Optional[str] = None, timeout: Optional[int] = 0) -> Path`**

Download the model file located at source to the models cache and return its Path. This will check `models/.download_cache` for the desired model file and download it from the indicated source if not already present. The local Path to the downloaded file is then returned.

---

## Other Changes

This PR performs a migration, in which it renames `models/.cache` to `models/.convert_cache`, and migrates previously-downloaded ESRGAN, openpose, DepthAnything and Lama inpaint models from the `models/core` directory into `models/.download_cache`. 

There are a number of legacy model files in `models/core`, such as GFPGAN, which are no longer used. This PR deletes them and tidies up the `models/core` directory. 

## Related Issues / Discussions

I have  systematically replaced all the calls to `download_with_progress_bar()`. This function is no longer used elsewhere and has been removed.

<!--WHEN APPLICABLE: List any related issues or discussions on github or discord. If this PR closes an issue, please use the "Closes #1234" format, so that the issue will be automatically closed when the PR merges.-->

## QA Instructions

I have added unit tests for the three new calls. You may test that the `load_and_cache_model()` call is working by running the upscaler within the web app. On first try, you will see the model file being downloaded into the models `.cache` directory. On subsequent tries, the model will either load from RAM (if it hasn't been displaced) or will be loaded from the filesystem.

<!--WHEN APPLICABLE: Describe how we can test the changes in this PR.-->

## Merge Plan

Squash merge when approved.

<!--WHEN APPLICABLE: Large PRs, or PRs that touch sensitive things like DB schemas, may need some care when merging. For example, a careful rebase by the change author, timing to not interfere with a pending release, or a message to contributors on discord after merging.-->

## Checklist

- [X] _The PR has a short but descriptive title, suitable for a changelog_
- [X] _Tests added / updated (if applicable)_
- [X] _Documentation added / updated (if applicable)_

lstein commented:
There are two issues with the current code. The first is that the models download to the `.cache` directory and may be cleared periodically by the model manager in which case they will be re-downloaded when next needed. I’m not sure whether this is a feature or a bug, but I’m going to change the code so that models download somewhere that isn’t volatile.

The second issue is that if a model is already installed (usually in `models/core`), the existing copy is ignored. I’ll add a migrate script.

psychedelicious commented on invokeai/app/invocations/upscale.py:
Should we use the loader context manager here, and remove the call at the end to empty the torch cache?

lstein replied:
Good idea! I should have thought to do that.
Now done.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -426,10 +430,161 @@ def search_by_attrs(
+        )
+        return path
+
+    def load_ckpt_from_url(
+        self,
+        source: Union[str, AnyHttpUrl],
+        access_token: Optional[str] = None,
+        timeout: Optional[int] = 0,
+        loader: Optional[Callable[[Path], Dict[str | int, Any]]] = None,
+    ) -> LoadedModel:
```
`load_ckpt_from_url(...)` currently calls `download_and_cache_ckpt(...)` internally. What do you think of this alternative split?:
- `def download_and_cache_ckpt(self, source: Union[str, AnyHttpUrl]) -> Path`
    - No change from what it does now.
- `def load_custom_model(self, path: Path) -> LoadedModel`
    - Loads a model from a path and registers it with the model cache.
    - Raises if the path does not exist.

This gives better separation of concerns and gives more flexibility to the user. For example, it gives the ability for a community node to load custom models from disk without needing them to be publicly downloadable.

The one downside I can think of is that it would cause an extra check that the model file exists on disk in cases where the model is already in the cache. Maybe this is why you did it the other way to begin with? This is definitely solvable through caching if it actually begins a performance issue.

lstein replied:
The split is fine with me. How about we keep `load_ckpt_from_url()` as a three-liner that does:

```
model_path = download_and_cache_ckpt()
load_custom_model(model_path)
```

There is no need for an additional path exists check, as `download_and_cache_ckpt()` already does this, and will raise an exception if the file neither exists in the cache nor can be downloaded.

I haven't implemented this yet - let me know what you think.

RyanJDick replied:
> The split is fine with me. How about we keep load_ckpt_from_url() as a two-liner that does:

Sure, that sounds good to me.

> There is no need for an additional path exists check, as download_and_cache_ckpt() already does this, and will raise an exception if the file neither exists in the cache nor can be downloaded.

Right. What I meant was that by splitting it up, we end up doing this check every time. Whereas with the current `load_ckpt_from_url(...)` it can be skipped if the model is in the cache.

lstein replied:
Ok, done.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -426,10 +430,161 @@ def search_by_attrs(
+                else lambda path: safetensors_load_file(path, device="cpu")
+            )
+
+        raw_model = loader(path)
+        ram_cache.put(key=str(source), model=raw_model)
+        return LoadedModel(_locker=ram_cache.get(key=str(source)))
```
This loader pattern is clever. I think this is the first time we've enabled community nodes to register custom models with the model cache, right? That's pretty cool!

At some point we may want to formalize some sort of `CacheableModel` ABC interface that all models registered with the model cache have to implement. But, I think that can be left for future work.

lstein replied:
Nice idea. I'll make a stab at a `CacheableModel` in some future PR (if nobody does it first).

lstein replied:
On my list to do.

psychedelicious commented on invokeai/backend/image_util/infill_methods/lama.py:
```diff
@@ -28,18 +25,14 @@ def load_jit_model(url_or_path, device):
-                url="https://github.com/Sanster/models/releases/download/add_big_lama/big-lama.pt",
-                dest_path=model_location,
-            )
+    def __init__(self, context: InvocationContext):
```
We shouldn't be passing the context into these classes, it makes them unusable outside nodes.

lstein replied:
Ok. I'll have to refactor `invocation_context.py` and move `load_ckpt_from_url()` into the model manager service. It probably belongs there anyway. Getting late, so leaving till tomorrow.

lstein replied:
Done now, but I had to figure out a way to pass the currently active model manager service to the LaMA constructor, and ended up doing this in `invocations/infill.py`:

```
lama = LaMA(context._services.model_manager)
```

I'd prefer not to access the protected `_services` attribute. Is there a better way to do this?

lstein commented on invokeai/app/invocations/controlnet_image_processors.py:
```diff
@@ -291,7 +291,8 @@ class MidasDepthImageProcessorInvocation(ImageProcessorInvocation):
 
-    def run_processor(self, image):
+    def run_processor(self, image: Image.Image, context: InvocationContext) -> Image.Image:
+        # TODO: replace from_pretrained() calls with context.models.download_and_cache() (or similar)
```
Happy to {help with / do this} after this PR is cleaned up and goes in. Let me know.

psychedelicious replied:
That would be great.

I snooped around a bit more and realized that the processor classes are mostly copy-pastes from papers that implement the processors. The controlnet author only added some posturing related to loading models and resizing the images.

For example, the midas depth map generation logic was copied from a reference implementation, to the OG controlnet repo + the auxiliary logic, and then finally to `controlnet_aux`.

Explains why there are a number of very different code styles in there.

lstein replied:
I'll start working on this sometime in the coming week.

lstein commented on invokeai/app/services/model_install/model_install_default.py:
```diff
@@ -394,15 +394,19 @@ def unconditionally_delete(self, key: str) -> None:  # noqa D102
             rmtree(model_path)
         self.unregister(key)
 
+    @classmethod
+    def _download_cache_path(cls, source: Union[str, AnyHttpUrl], app_config: InvokeAIAppConfig) -> Path:
+        model_hash = sha256(str(source).encode("utf-8")).hexdigest()[0:32]
+        return app_config.download_cache_path / model_hash
+
```
Done. The directories are now a properly escaped version of the HTTP URL, e.g. `https__githubcom_xinntao_real-esrgan_releases_download_v010_realesrgan_x4pluspth`

I assume you also would like the converted safetensors models in the convert cache to have more informative names. I've changed them to look like this: 
`home_lstein_invokeai-main_models_sdxl_main_stable_diffusion_1_0`. I don't much care for how these look, but the easiest way to ensure the names were unique was to use the absolute path.

psychedelicious replied:
That's great, thanks.

lstein commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -426,10 +430,161 @@ def search_by_attrs(
             model_format=format,
         )
 
+    def install_model(
```
Removed. I hope I can find it again if similar functionality is needed in the future.

lstein commented:
@psychedelicious @RyanJDick I think I've responded to all comments and suggestions. Thanks for the reviews!

psychedelicious reviewed (changes requested):
I'll do a full review later, we still have too much coupling between the context/model manager and some of the utility classes.

I've just pushed changes that lift the loading out of the LaMa and DepthAnything classes, but not DWOpenPose - I'm not familiar with how ONNX loads models and ran out of time for this for now..

lstein commented:
> I'll do a full review later, we still have too much coupling between the context/model manager and some of the utility classes.
> 
> I've just pushed changes that lift the loading out of the LaMa and DepthAnything classes, but not DWOpenPose - I'm not familiar with how ONNX loads models and ran out of time for this for now..

There just doesn't seem to be a good way for utility classes to access the model manager service in the running app without going through the context. One out would be to have the model manager set a singleton in the way that the configuration service does, thereby allowing the utility classes make a call to `get_model_manager()`. Of course, there would have to be a bit of setup in the event that the model manager hadn't already been initialized.

psychedelicious commented:
> There just doesn't seem to be a good way for utility classes to access the model manager service in the running app without going through the context. One out would be to have the model manager set a singleton in the way that the configuration service does, thereby allowing the utility classes make a call to get_model_manager(). Of course, there would have to be a bit of setup in the event that the model manager hadn't already been initialized.

Do my changes in these two commits not do this?
[feat(backend): lift managed model loading out of lama class](https://github.com/invoke-ai/InvokeAI/pull/6132/commits/fcb071f30caa5de52ab6e951e39dfbf6bf80e458)
[feat(backend): lift managed model loading out of depthanything class](https://github.com/invoke-ai/InvokeAI/pull/6132/commits/1fe90c357cb3f216bd4f2789180911c3520c6505)

lstein commented:
> > There just doesn't seem to be a good way for utility classes to access the model manager service in the running app without going through the context. One out would be to have the model manager set a singleton in the way that the configuration service does, thereby allowing the utility classes make a call to get_model_manager(). Of course, there would have to be a bit of setup in the event that the model manager hadn't already been initialized.
> 
> Do my changes in these two commits not do this? [feat(backend): lift managed model loading out of lama class](https://github.com/invoke-ai/InvokeAI/pull/6132/commits/fcb071f30caa5de52ab6e951e39dfbf6bf80e458) [feat(backend): lift managed model loading out of depthanything class](https://github.com/invoke-ai/InvokeAI/pull/6132/commits/1fe90c357cb3f216bd4f2789180911c3520c6505)

Sure. Those look good.

psychedelicious commented:
Sorry, no we need to do the same pattern for that last processor, `DWOpenPose` - it uses ONNX models and I'm not familiar with how they work. I can take care of that but won't be til next week.

lstein commented:
> Sorry, no we need to do the same pattern for that last processor, `DWOpenPose` - it uses ONNX models and I'm not familiar with how they work. I can take care of that but won't be til next week.

I'll take a look at it Friday.

lstein commented:
I've refactored `DWOpenPose` using the same pattern as in the other backend image processors. I also added some of the missing typehints so there are fewer red squigglies.  I noticed that there is a problem with the pip dependencies. If the `onnxruntime` package is installed, then even if `onnxruntime-gpu` is installed as well, the onnx runtime won't use the GPU (see https://github.com/microsoft/onnxruntime/issues/7748). You have to remove `onnxruntime` and then install `onnxruntime-gpu`. I don't think `pyproject.toml` provides a way for an optional dependency to remove a default dependency. Is there a workaround?

psychedelicious commented:
> I noticed that there is a problem with the pip dependencies. If the onnxruntime package is installed, then even if onnxruntime-gpu is installed as well, the onnx runtime won't use the GPU (see https://github.com/microsoft/onnxruntime/issues/7748). You have to remove onnxruntime and then install onnxruntime-gpu. I don't think pyproject.toml provides a way for an optional dependency to remove a default dependency. Is there a workaround?

I think we'd need to just update the installer script with special handling to uninstall those packages if they are already installed. It's probably time to revise our optional dependency lists. I think "cuda" and "cpu" make sense to be the only two user-facing options. "xformers" is extraneous now (torch's native SDP implementation is just as fast), so it could be removed.

psychedelicious commented:
Thanks for cleaning up the pose detector. It would be nice to use the model context so we get memory mgmt, but that is a future task.

I had some feedback from earlier about the public API that I think was lost:

> 1. Token: When would a node reasonably provide an API token? We support regex-matched tokens in the config file. I don't think this should be in the invocation API.
> 
> 2. Timeout: Similarly, when could a node possibly be able to make a good determination of the timeout for a download? It doesn't know the user's internet connection speed. It's a user setting - could be globally set in the config file and apply to all downloads.

If both of those args are removed, then `load_ckpt_from_path` and `load_ckpt_from_url` look very similar. I think this is maybe what @RyanJDick was suggesting with a single `load_custom_model` method.

Also, will these methods work for diffusers models? If so, "ckpt" probably doesn't need to be in the name.

lstein commented:
> Thanks for cleaning up the pose detector. It would be nice to use the model context so we get memory mgmt, but that is a future task.
> 
> 
> 
> I had some feedback from earlier about the public API that I think was lost:
> 
> 
> 
> > 1. Token: When would a node reasonably provide an API token? We support regex-matched tokens in the config file. I don't think this should be in the invocation API.
> 
> > 
> 
> > 2. Timeout: Similarly, when could a node possibly be able to make a good determination of the timeout for a download? It doesn't know the user's internet connection speed. It's a user setting - could be globally set in the config file and apply to all downloads.
> 
> 
> 
> If both of those args are removed, then `load_ckpt_from_path` and `load_ckpt_from_url` look very similar. I think this is maybe what @RyanJDick was suggesting with a single `load_custom_model` method.
> 
> 
> 
> Also, will these methods work for diffusers models? If so, "ckpt" probably doesn't need to be in the name.

I'll remove the arguments and consolidate the two calls. Diffusers loading is NOT supported right now, so ckpt needs to stay in the names.

psychedelicious commented:
Probably doesn't make sense to spend time on the onnx loading. This is the only model that uses it.

> Right now the regex token handling is done in a part of the install manager that is not called by the simplifed API. I'll move this code into the core download() routine so that tokens are picked up whenever a URL is requested.

Sounds good.

> I think you're saying this should be a global config option and I agree with that. Can we get the config migration code in so that I have a clean way of updating the config?

I don't think any migration is necessary - just add a sensible default value. I'll check back in on the config migration PR this week.

> Not currently. It only works with checkpoints. I'd planned to add diffusers support later, but I guess I should do that now. Converting to draft.

Ok, thanks.

lstein commented:
> The onnxruntime model loading architecture seems to be very different from what the model manager expects. In particular, the `onnxruntime.InferenceSession()` constructor doesn't seem to provide any way to accept a model that has been read into RAM or VRAM. The closest I can figure is that you can pass the constructor an `IOBytes` object to a serialized version of the model in memory. This will require some architectural changes in the model manager that should be its own PR.

I've played with this a bit. It is easy to load the openpose onnx sessions into the RAM cache and they will run happily under the existing MM cache system. However, Onnx sessions do their own internal VRAM/CUDA management, and so I found that for the duration of the time that the session object is in RAM, it holds on to a substantial chunk of VRAM (1.7GB). The openpose session is only used during conversion of an image into a pose model, and I think it's better to have slow disk-based loading of the openpose session than to silently consume a chunk of VRAM that interferes with later generation.

lstein commented:
@psychedelicious This is ready for your review now. There are now just two calls: `load_and_cache_model()` and `download_and_cache_model()` which return a locally cached Path and LoadedModel respectively. In addition, the model source can now be a URL, a local Path, or a repo_id. Support for the latter involved my refactoring the way that multifile downloads work.

lstein commented:
@psychedelicious I just updated the whole thing to work properly with the new (and very nice) Pydantic-based events. I've also added a new migration. Please review when you can. This merge took me over an hour to resolve and I'd like to avoid repeating the experience.

psychedelicious commented on invokeai/backend/model_manager/load/load_base.py:
```diff
@@ -23,8 +23,8 @@
     """Context manager object that mediates transfer from RAM<->VRAM."""
 
-    config: AnyModelConfig
     _locker: ModelLockerBase
+    config: Optional[AnyModelConfig] = None
```
This is a notable API change. This means we cannot rely on a LoadedModel to have a config. Maybe this should be two separate classes, `LoadedModelWithoutConfig` and `LoadedModel(LoadedModelWithoutConfig)`...

lstein replied:
I did exactly this. Thanks for the pointer.

psychedelicious commented:
I removed a number of unnecessary changes in `invocation_context.py`, mostly extraneous type annotations. If mypy is complaining about these, then that's a mypy problem, because all the methods are annotated correctly.

psychedelicious commented on invokeai/app/services/model_load/model_load_default.py:
```diff
@@ -75,3 +83,41 @@ def load_model(self, model_config: AnyModelConfig, submodel_type: Optional[SubMo
+            result: Dict[str, Tensor] = torch_load(checkpoint, map_location="cpu")
+            return result
+
+        def diffusers_load_directory(directory: Path) -> AnyModel:
+            load_class = GenericDiffusersLoader(
+                app_config=self._app_config,
+                logger=self._logger,
+                ram_cache=self._ram_cache,
+                convert_cache=self.convert_cache,
+            ).get_hf_load_class(directory)
+            result: AnyModel = load_class.from_pretrained(model_path, torch_dtype=TorchDevice.choose_torch_dtype())
+            return result
```
This function is unused - I think the logic to get the loader should be checking if it's a directory? I'm not sure how to fix this myself bc the `diffusers_load_directory` function has a different type signature than the other loader function options.

lstein replied:
Fixed. Looks like I never wired up that function! In the process I discovered a long-standing bug that would prevent text encoders from being loaded generically. 

I also added a unit test for loading generic diffusers models. To do this, I added the 10 MB "tiny" taesdxl model, which was the smallest loadable diffusers I could find.

psychedelicious replied:
I don't think we should be adding models to the repo. Feels arbitrary to do it just for this one scenario and not all of them, plus 10MB increases the repo size by like 8%. Don't wanna get in the habit of doing this.

I think the best way to test model loading is to package up a models dir that has samples of various model types and configurations, then run full loading tests locally on one of our own machines or a custom CI runner.

lstein replied:
I've commented out the unit test for directory loading.

psychedelicious replied:
Can you please drop the commits that added the model file so it's not in the repo?

psychedelicious replied:
I've removed the model files from git history and force-pushed

lstein commented:
@psychedelicious I've addressed the remaining issues you raised. Thanks for a thorough review.

lstein commented:
> I removed a number of unnecessary changes in `invocation_context.py`, mostly extraneous type annotations. If mypy is complaining about these, then that's a mypy problem, because all the methods are annotated correctly.
> 
> I also moved `load_model_from_url` from the main model manager class into the invocation context.

Yes, mypy is having trouble tracking the return type of several methods. I haven't figured out what causes the problem and don't want to add a # type: ignore. But maybe I should 'cause I'm not ready to turn to pyright.

psychedelicious commented:
> Yes, mypy is having trouble tracking the return type of several methods. I haven't figured out what causes the problem and don't want to add a # type: ignore. But maybe I should 'cause I'm not ready to turn to pyright.

We shouldn't add `# type: ignore`, that will stop all type checkers from doing anything - including pyright. The places where you made code quality concessions to satisfy mypy involve very straightforward types - either your mypy config is borked or mypy itself is borked. FWIW, I've found pyright to be much faster, more thorough and more correct than mypy.

psychedelicious commented:
@RyanJDick Would you mind doing one last review of this PR?

lstein commented:
> > Yes, mypy is having trouble tracking the return type of several methods. I haven't figured out what causes the problem and don't want to add a # type: ignore. But maybe I should 'cause I'm not ready to turn to pyright.
> 
> We shouldn't add `# type: ignore`, that will stop all type checkers from doing anything - including pyright. The places where you made code quality concessions to satisfy mypy involve very straightforward types - either your mypy config is borked or mypy itself is borked. FWIW, I've found pyright to be much faster, more thorough and more correct than mypy.

You've convinced me. I've switched to pyright!

RyanJDick commented:
> @RyanJDick Would you mind doing one last review of this PR?

Looks like 43/44 files have changed since I last looked 😅 . I'll plan to spend a chunk of time on this tomorrow.

psychedelicious commented:
@RyanJDick Can narrow that down to reviewing `invocation_context.py`, which changes the public API and is more important to get right the first time. Thanks.

RyanJDick commented on invokeai/app/services/shared/sqlite_migrator/migrations/migration_11.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def build_migration_11(app_config: InvokeAIAppConfig, logger: Logger) -> Migration:
+    """
+    Build the migration from database version 9 to 10.
+
+    This migration does the following:
+    - Moves "core" models previously downloaded with download_with_progress_bar() into new
+      "models/.download_cache" directory.
+    - Renames "models/.cache" to "models/.convert_cache".
+    - Adds `error_type` and `error_message` columns to the session queue table.
+    - Renames the `error` column to `error_traceback`.
+    """
```
This docstring is outdated.

lstein replied:
Fixed.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -434,6 +437,55 @@ def search_by_attrs(
+            Path to the downloaded model
+        """
+
+        return self._services.model_manager.install.download_and_cache_model(source=source)
```
There is a type mismatch between these source args: `str` vs. `str  | AnyHttpUrl`

lstein replied:
Fixed.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -434,6 +437,55 @@ def search_by_attrs(
+            A LoadedModelWithoutConfig object.
+        """
+
+        if isinstance(source, Path):
+            return self._services.model_manager.load.load_model_from_path(model_path=source, loader=loader)
+        else:
+            model_path = self._services.model_manager.install.download_and_cache_model(source=str(source))
+            return self._services.model_manager.load.load_model_from_path(model_path=model_path, loader=loader)
```
I don't love that we switch behaviour based on whether `source` is a `Path` or a `str`. It feels like a fragile distinction, especially given the popularity of using `str`s to represent paths in Python. The caller should always know whether they are dealing with a path or a URL/repo name, so I think it's better to make this distinction explicit.

In [this discussion](https://github.com/invoke-ai/InvokeAI/pull/6132#discussion_r1576894928) we had landed on an API that didn't require this type condition. Was there a reason for moving away from that?

lstein replied:
It looks like @psychedelicious removed `load_model_from_url()` in commit b12444 . Add it back?

psychedelicious replied:
I moved the method from the high-level `services.model_manager` class to the invocation context. The function is otherwise the same.

lstein replied:
@psychedelicious  I can’t find `load_model_from_url()` in `invocation_context.py`, and the commit indicates that it was deleted from the model manager and then the `load_and_cache_model()` was modified to no longer call it.

I’ll take care of fixing the API if the change wasn’t deliberate.

lstein replied:
For clarity, I have renamed the methods `load_local_model()` and `load_remote_model()`. The former accepts the Path to a file or directory, and the latter accepts either a direct download URL or a HuggingFace URL. I have fixed the documentation and updated the pull request description.

psychedelicious replied:
Sounds good!

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -434,6 +437,55 @@ def search_by_attrs(
+    def load_and_cache_model(
+        self,
+        source: Path | str | AnyHttpUrl,
+        loader: Optional[Callable[[Path], dict[str, Tensor]]] = None,
```
The current loader typing doesn't match how it is being used. One of these options might be better:
```python
        loader: Optional[Callable[[Path], Any]] = None,
```
```python
        loader: Optional[Callable[[Path], torch.nn.Module]] = None,
```

lstein replied:
Fixed.

RyanJDick reviewed (commented):
I just reviewed the `invocation_context.py` API.

lstein commented:
@RyanJDick I've fixed the issues you identified.

RyanJDick reviewed (commented):
`invocation_context.py` looks good to me ✅ 

I'll defer to @psychedelicious for final approval, since he has a more complete understanding of this PR than me at this point.
